src/picongpu/include/particles/ionization/byField/BSI/BSI_Impl.hpp

```diff
@@ -0,0 +1,195 @@
+#include "fields/FieldE.hpp"
+
+#include "particles/ionization/byField/BSI/BSI.def"
+#include "particles/ionization/byField/BSI/AlgorithmBSI.hpp"
+#include "particles/ionization/ionization.hpp"
+
+#include "compileTime/conversion/TypeToPointerPair.hpp"
+#include "memory/boxes/DataBox.hpp"
+
+#include "particles/ParticlesFunctors.hpp"
+
+namespace picongpu
+{
+namespace particles
+{
+namespace ionization
+{
+
+    /** \struct BSI_Impl
+     * \brief Barrier Suppression Ionization - Implementation
+     *
+     * \tparam T_DestSpecies electron species to be created
+     * \tparam T_SrcSpecies particle species that is ionized
+     */
+    template<typename T_DestSpecies, typename T_SrcSpecies>
+    struct BSI_Impl
+    {
+
+        typedef T_DestSpecies DestSpecies;
+        typedef T_SrcSpecies  SrcSpecies;
+
+        typedef typename SrcSpecies::FrameType FrameType;
+
+        /* specify field to particle interpolation scheme */
+        typedef typename PMacc::traits::Resolve<
+            typename GetFlagType<FrameType,interpolation<> >::type
+        >::type InterpolationScheme;
+
+        /* margins around the supercell for the interpolation of the field on the cells */
+        typedef typename GetMargin<InterpolationScheme>::LowerMargin LowerMargin;
+        typedef typename GetMargin<InterpolationScheme>::UpperMargin UpperMargin;
+
+        /* relevant area of a block */
+        typedef SuperCellDescription<
+            typename MappingDesc::SuperCellSize,
+            LowerMargin,
+            UpperMargin
+            > BlockArea;
+
+        BlockArea BlockDescription;
+
+        private:
+
+            /* define ionization ALGORITHM (calculation) for ionization MODEL */
+            typedef particles::ionization::AlgorithmBSI IonizationAlgorithm;
+
+            typedef MappingDesc::SuperCellSize TVec;
+
+            typedef FieldE::ValueType ValueType_E;
+            typedef FieldB::ValueType ValueType_B;
+            /* global memory EM-field device databoxes */
+            FieldE::DataBoxType eBox;
+            FieldB::DataBoxType bBox;
+            /* shared memory EM-field device databoxes */
+            PMACC_ALIGN(cachedE, DataBox<SharedBox<ValueType_E, typename BlockArea::FullSuperCellSize,1> >);
+            PMACC_ALIGN(cachedB, DataBox<SharedBox<ValueType_B, typename BlockArea::FullSuperCellSize,0> >);
+
+        public:
+            /* host constructor */
+            BSI_Impl(const uint32_t currentStep)
+            {
+                DataConnector &dc = Environment<>::get().DataConnector();
+                /* initialize pointers on host-side E-(B-)field databoxes */
+                FieldE* fieldE = &(dc.getData<FieldE > (FieldE::getName(), true));
+                FieldB* fieldB = &(dc.getData<FieldB > (FieldB::getName(), true));
+                /* initialize device-side E-(B-)field databoxes */
+                eBox = fieldE->getDeviceDataBox();
+                bBox = fieldB->getDeviceDataBox();
```

Comment:
I bet if we did a **PIConGPU documentation workshop** many opportunities for refactoring would become clear. :stuck_out_tongue: